### PR TITLE
07 virtual environments to iterate and test

### PR DESCRIPTION
**Pull request by gregordecristoforo**

### CODE REVIEW

**Comment by Chroxvi** [comment]
A few things, you may want to consider:

- This venv approach may also be used with the LUMI application containers, e.g. `/appl/local/containers/sif-images/lumi-pytorch-rocm-5.6.1-python-3.10-pytorch-v2.2.2.sif`, which may be worth mentioning. It still requires activating the conda environment (`$WITH_CONDA`) before creating the venv. Also note that building a (final) container from e.g. `/appl/local/containers/sif-images/lumi-pytorch-rocm-5.6.1-python-3.10-pytorch-v2.2.2.sif` + a venv is not necessarily that easy - at least it is not directly supported by e.g. cotainr.
- I think, we should stress that you need to source the venv activate script every time you run the container to get access to the venv installed packages (after running `$WITH_CONDA` if using the LUMI application containers). And you need to keep track of which venv to activate with a given container for a given application/project (another reason to build a container with everything included, or, alternatively, use the EasyBuild integration to the LUMI Application containers to manage this).
- I also think, we should stress that the venv is installed on the Lustre files systems (not inside the container which is read-only) with all the small files / bad performance issues it may create.
- Torchvision is actually already installed in the `minimal_python.yml` conda env. It makes it slightly confusing to install the same package again in a venv. Maybe choose another package to install instead?
- The `singularity-userfilesystems` module is not (yet) available in the central LUMI stacks. You might want to mention that `module use /project/project_465001063/modules` is currently needed to have access to it.

**Comment by gregordecristoforo** [comment]
Thank you @Chroxvi  for your comments, I incorporated them in the guide. (Sorry for mentioning the wrong package that is installed via pip, I intended to install `torchmetrics` instead of `torchvision` which was already correct in the `pip list` but not in the explicit `pip install`).
Are there any other issues/comments before it can be merged?

**Comment by Chroxvi** [comment]
Since we no longer have a hands-on session for this lecture, I wonder if you rework this into a "minimal full example" and then keep the remarks and discussion in your slides? What do you think?

**Comment by gregordecristoforo** [comment]
I turned the relatively detailed instructions into a "minimal full example". All the removed details are still included in the slides.

File: 07_Virtual_environments_to_iterate_and_test/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+## Set up virtual environment
+
+First we run a shell inside the container
+```bash
+singularity shell --bind /pfs,/scratch,/projappl,/project,/flash,/appl minimal_pytorch.sif
+```
+Note that setting `--bind` is optional, you achieve the same by
+```bash
+module use /project/project_465001063/modules
```
Comment (Chroxvi) [inline 07_Virtual_environments_to_iterate_and_test/README.md]:
```suggestion
module use /appl/local/training/modules/AI-20240529/
```

File: 07_Virtual_environments_to_iterate_and_test/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+```
+The `--system-site-packages` flag gives the virtual environment access to the packages from the container.
+
+## Install custom packages
+
+After activating the virtual environment we can now install custom packages via pip, for example:
+```bash
+pip install torchmetrics
+```
+
```
Comment (Chroxvi) [inline 07_Virtual_environments_to_iterate_and_test/README.md]: Maybe add a section showing how to correctly run the container with the venv, i.e. sourcing the venv.

**Review by Chroxvi — CHANGES_REQUESTED** [review]
Generally looks good. The custom modules path needs to be updated. And maybe add a short description of how to run the container with the venv. See comments.

**Comment by gregordecristoforo** [comment]
Thanks for the comments, I included your suggestions.

**Review by Chroxvi — APPROVED** [review]
Ready for merge!

**Pull request merged**